src/pybind/mgr/cephadm/module.py

```diff
@@ -2980,7 +2980,15 @@ def upgrade_check(self, image: str, version: str) -> str:
         }
         for host, dm in self.cache.daemons.items():
             for name, dd in dm.items():
-                if image_info.image_id == dd.container_image_id:
+                # check if the container digest for the digest we're checking upgrades for matches
+                # the container digests for the daemon if "use_repo_digest" setting is true
+                # or that the image name matches the daemon's image name if "use_repo_digest"
+                # is false. The idea is to generally check if the daemon is already using
+                # the image we're checking upgrade to.
+                if (
+                    (self.use_repo_digest and dd.matches_digests(image_info.repo_digests or []))
```

Comment:
After your last changes, do we still need the `or []` in the call? .

.. the new code seems to already check for an empty list / None value so it seems that we should be OK:

```
if not digests or not self.container_image_digests:
            return False
```

Reply:
the type signature on the `matches_digest` function as it is now expects the digests to be type `List[str]` so that would cause mypy errors. What I could do is go change that to make the type `Optional` and then come back and remove the `or []`